### PR TITLE
Loading old model files is broken.

### LINKED ISSUE #1289: Exception loading an older model file with no strings
When I changed our model format for the ComponentCatalog changes (#970), I refactored some model loading code incorrectly. There are some existing models that are breaking our validation checks:

The old code was doing:

https://github.com/dotnet/machinelearning/blob/655c2e29421f5c44133ed6515374e48cf42868a2/src/Microsoft.ML.Data/Model/ModelHeader.cs#L419-L424

The new code now does:

https://github.com/dotnet/machinelearning/blob/a44e131b8010024d024197db4cc22a546c554fdf/src/Microsoft.ML.Data/Model/ModelHeader.cs#L482-L486

Notice that it doesn't return early in the new code. Lower in the method there are checks here:

https://github.com/dotnet/machinelearning/blob/a44e131b8010024d024197db4cc22a546c554fdf/src/Microsoft.ML.Data/Model/ModelHeader.cs#L538-L541

These checks are now failing with older model files. They were never run in the old code when there were no strings, so we shouldn't be running them anymore when reading older model files.

/cc @yaeldekel

### PR DESCRIPTION
**Pull request by eerhardt**
Older model files are failing to load due to FpTail checks in the model validation code. These checks weren't happening in the old model loading code, and they now fail on some older models.

Fix this by returning early when it is an older model, and there are no strings. This preserves the old behavior.

Fix #1289

### CODE REVIEW

File: test/Microsoft.ML.Core.Tests/UnitTests/TestModelLoad.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        /// <summary>
+        /// Tests loading a model file that was saved using an older version still loads correctly.
+        /// </summary>
+        [Fact]
+        public void LoadOriginalBinaryLoaderModel()
+        {
+            using (var env = new LocalEnvironment()
+                .AddStandardComponents())
+            using (var modelStream = File.OpenRead("OriginalVersionBinaryLoader.zip"))
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Core.Tests/UnitTests/TestModelLoad.cs]: >OriginalVersionBinaryLoader [](start = 52, length = 27)

Do we have this file in repo?

Comment (eerhardt) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestModelLoad.cs]: It comes through the TestModels NuGet package I added.

File: build/Dependencies.props [hunk 2]
```diff
@@ -1,24 +1,44 @@
...
-    <SystemThreadingTasksDataflowPackageVersion>4.8.0</SystemThreadingTasksDataflowPackageVersion>
     <SystemCodeDomPackageVersion>4.4.0</SystemCodeDomPackageVersion>
+    <SystemCollectionsImmutableVersion>1.5.0</SystemCollectionsImmutableVersion>
+    <SystemMemoryVersion>4.5.1</SystemMemoryVersion>
     <SystemReflectionEmitLightweightPackageVersion>4.3.0</SystemReflectionEmitLightweightPackageVersion>
-    <PublishSymbolsPackageVersion>1.0.0-beta-62824-02</PublishSymbolsPackageVersion>
+    <SystemThreadingTasksDataflowPackageVersion>4.8.0</SystemThreadingTasksDataflowPackageVersion>
+  </PropertyGroup>
+
+  <!-- Leaf Project Product Dependencies -->
```
Comment (Ivanidzo4ka) [inline build/Dependencies.props]: >Leaf Project [](start = 7, length = 12)

Is it common terminology for projects you depend on?

Comment (eerhardt) [reply build/Dependencies.props]: I just meant “non-core” product dependencies. I heard them called “leaf projects” the other day, meaning no other projects depend on them directly.

I can use different wording, if there is some better terminology.

File: test/Microsoft.ML.Core.Tests/UnitTests/TestModelLoad.cs (hunk 1 above)
Comment (Zruty0) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestModelLoad.cs]: Is this sustainable? Suppose Scott wants to add a couple more models to test back-compat. How can he do this?

---
In reply to: [226122081](https://github.com/dotnet/machinelearning/pull/1290#discussion_r226122081) [](ancestors = 226122081)

Comment (eerhardt) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestModelLoad.cs]: Yes, it is sustainable. I am following the same pattern we use for ONNX and TensorFlow test models.

You check the model file into a folder in https://github.com/dotnet/machinelearning-testdata. For example, if Scott has more ML.NET model files, he adds them under https://github.com/dotnet/machinelearning-testdata/tree/master/Microsoft.ML.TestModels, sends a PR and merges. Then someone builds the NuGet package from that repo, and uploads it to our NuGet feed: https://dotnet.myget.org/feed/dotnet-core/package/nuget/Microsoft.ML.TestModels.

The NuGet packages are versioned, so as new versions are created (with new or updated model files), we can just bump the version number in this repo, and start using the new models in the test code.

Comment (justinormont) [inline test/Microsoft.ML.Core.Tests/UnitTests/TestModelLoad.cs]: I might recommend naming this model w/ the version number which created it. Saying 'OriginalVersion' makes we think we'll have a plethora of non-comparable names like 'NewVersion', and 'NewVersionAfterUpdate'. 

```suggestion
            using (var modelStream = File.OpenRead("BinaryLoader-v3.11.0.0.zip"))
```

Comment (eerhardt) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestModelLoad.cs]: Good idea. I need to update the TestModels NuGet package, which I'm doing here: https://github.com/dotnet/machinelearning-testdata/pull/8/.

Once that is approved, merged, and published, I will update this PR.

**Comment by justinormont** [comment]
Closing & reopening to restart CI tests.

Failure spot was on a nuget download for OSX both debug & release:
```
2018-10-18T16:37:41.1798700Z   Failed to download package 'Microsoft.ML.Scoring.1.1.0' from 'https://api.nuget.org/v3-flatcontainer/microsoft.ml.scoring/1.1.0/microsoft.ml.scoring.1.1.0.nupkg'.
2018-10-18T16:37:41.1799680Z   The download of 'https://api.nuget.org/v3-flatcontainer/microsoft.ml.scoring/1.1.0/microsoft.ml.scoring.1.1.0.nupkg' timed out because no data was received for 60000ms.
2018-10-18T16:37:41.1800340Z     Exception of type 'System.TimeoutException' was thrown.
```

**Pull request merged**